### DISCUSSION TITLE
Imports and exports for external functions

### DISCUSSION THREAD

**Discussion by BobVarioa** [CanadaHonk/porffor#165] [Ideas]
For #131, we likely want to add support for handling external imports and exports. In my mind, any export statements should be exported in Wasm as well.
But, for import statements it's less cut and dry, my main thought was to use a special `wasm:imports` url which could import anything that the host environment defined.
Something like this: 
```js
import { foo } from "wasm:imports";

foo(1);

export const bar = () => 2;
```

But, do we want first class support for any language's files? My immediate thought is C, because of `2c` but shipping a C parser and C preprocessor seems a little much for the project.
Also, what should the user interface look like for this? How should someone define Porffor imports? An IDL? Some JS interface? Etc.

**Comment by Rob23oba** [comment]
Hmm, just another idea to put out there: maybe you could import wasm functions using a special `Porffor.importFunction`, maybe similar to `Porffor.dlopen` but maybe also like this:
```js
const foo = Porffor.importFunction('example', 'foo', (i32, f64) => [ i32, i32 ]);
const [ x, y ] = foo(2, 3.5);
```
So essentially write a "template" function that tells porffor how the import should look like without requiring something like typescript. Even though this is kinda "syntax exploitation", I think this is an okay approach since there are other places where syntax has special meaning (e.g. `Porffor.wasm`) and the import signature is required to be known at compile time.
